### ISSUE TITLE
PNG thumbnails can't be generated for TIFFs with ICC profile warnings

### ISSUE THREAD

**Issue by AntiCompositeNumber** [libvips/libvips#1635]
If a TIFF file contains a problematic ICC profile, `vips shrink input.tiff output.png 20 20` will write an empty PNG file. JPEG generation, however, works without any errors or warnings. Originally reported on vips-8.4.5, confirmed on vips-8.10.0-Wed Apr 29 21:25:27 UTC 2020

Example files:
https://upload.wikimedia.org/wikipedia/commons/2/28/Julia_Margaret_Cameron_-_Queen_of_the_May_-_1984.166_-_Cleveland_Museum_of_Art.tif ([source](https://commons.wikimedia.org/wiki/File:Julia_Margaret_Cameron_-_Queen_of_the_May_-_1984.166_-_Cleveland_Museum_of_Art.tif))
```
$ vips shrink Julia_Margaret_Cameron_-_Queen_of_the_May_-_1984.166_-_Cleveland_Museum_of_Art.tif test.png 20 20

(vips:71759): VIPS-WARNING **: 12:44:43.543: profile 'icc': 0h: PCS illuminant is not D50
vips2png: unable to write to target
$ ls -hs test.png
0 test.png
$  identify -verbose The_Ramesêum._The_remains_of_the_great_hypostyle_of_the_Temple_of_Rameses_II._The_figures_of_Christian_slaves_have_been_painted_on_some_of_the_columns_\(NYPL_b10607452-80839\).tiff
[...]
    icc:copyright: Copyright (c) Eastman Kodak Company, 1999, all rights reserved.
    icc:description: ProPhoto RGB 
    icc:manufacturer: KODAK
    icc:model: Reference Output Medium Metric(ROMM)
[...]
```

https://upload.wikimedia.org/wikipedia/commons/archive/0/0a/20191126165135%21Moivre_-_Doctrine_of_chances%2C_1761_-_722666.tif ([source](https://commons.wikimedia.org/wiki/File:Moivre_-_Doctrine_of_chances,_1761_-_722666.tif))
```
$ vips shrink 20191126165135!Moivre_-_Doctrine_of_chances,_1761_-_722666.tif test2.png 20 20

(vips:74762): VIPS-WARNING **: 12:50:04.867: profile 'icc': 'desc': ICC profile tag start not a multiple of 4
vips2png: unable to write to target
$ ls -hs test2.png
0 test2.png
```

https://phabricator.wikimedia.org/T219569

**Comment by jcupitt** [comment]
Hello @AntiCompositeNumber, this is a recent libpng "feature". It will refuse to write an image if the ICC profile is inappropriate (eg. a CMYK profile on an RGB image).

I think I would transform all derived images to sRGB and strip the profile. Users can always download the original if they want the image as uploaded, and profiles can be many megabytes in extreme cases.

For example:

```
$ vipsthumbnail queen.tif --size 600x600 --output preview.png[strip] --export-profile srgb
```

**Comment by jcupitt** [comment]
You should also find that `vipsthumbnail` is faster and produces higher quality output than cooking something up yourself. There's a page on the libvips wiki which explains how `vipsthumbnail` works:

https://github.com/libvips/libvips/wiki/HOWTO----Image-shrinking

**Comment by AntiCompositeNumber** [comment]
Forcing non-sRGB images into the sRGB profile is less than ideal, as the preservation of color profiles in thumbnailing is expected by Wikimedia users. There is also an expectation that some metadata fields will be retained, so removing all metadata would be problematic (selective metadata stripping is implemented after vips). It would be best if it was possible to ignore the error and write the png anyway.

**Comment by jcupitt** [comment]
I agree, but unfortunately this is not possible, as far as I know. I think your options are:

- libpng will not write images with incompatible profiles;
- You could revert to an old libpng version, but of course that will expose you to a lot of security issues;
- There are several projects attempting to replace libpng, but none are production ready
- You could force all PNGs to sRGB;
- You could have a fallback path for png write which, on save failure, retries with a forced srgb conversion. This feels a bit hacky to me.

**Comment by kleisauke** [comment]
Commit https://github.com/libvips/libvips/commit/b9385b9d355ab64cb5e0314c00ccf535e265200b also fixes this, I think.

**Comment by jcupitt** [comment]
Yes, this should be fixed in 8.10. I'll close.
